src/t_set.c

```diff
@@ -525,15 +616,19 @@ void spopWithCountCommand(client *c) {
     if (remaining*SPOP_MOVE_STRATEGY_MUL > count) {
         while(count--) {
             /* Emit and remove. */
-            encoding = setTypeRandomElement(set,&sdsele,&llele);
+            encoding = setTypeRandomElement(set, &str, &len, &llele);
             if (encoding == OBJ_ENCODING_INTSET) {
                 addReplyBulkLongLong(c,llele);
                 objele = createStringObjectFromLongLong(llele);
                 set->ptr = intsetRemove(set->ptr,llele,NULL);
+            } else if (str == NULL) {
+                addReplyBulkLongLong(c,llele);
+                objele = createObject(OBJ_STRING, sdsfromlonglong(llele));
+                setTypeRemove(set, objele->ptr);
             } else {
-                addReplyBulkCBuffer(c,sdsele,sdslen(sdsele));
-                objele = createStringObject(sdsele,sdslen(sdsele));
-                setTypeRemove(set,sdsele);
+                addReplyBulkCBuffer(c, str, len);
+                objele = createStringObject(str, len);
+                setTypeRemove(set, objele->ptr);
             }
```

Comment:
Is it necessary to extract this code, because line 716 has the same code.
btw, I found that the lookup operation in `setTypeRemove` is actually wasteful, because `setTypeRandomElement` actually already knows the position of the lookedup element, but it looks it up again, maybe in the future we can get the iterator based on the index, and then delete by this iteraotr, so as to avoid these lookups.